icu4c/source/common/locid.cpp

```diff
@@ -504,34 +508,1071 @@ Locale::operator==( const   Locale& other) const
+    readScriptAlias(scriptAlias.getAlias(),
+                    strings.getAlias(),
+                    scriptTypes,
+                    scriptReplacementIndexes,
+                    scriptLength,
+                    status);
+
+    // Read the territoryAlias into territoryTypes, territoryReplacementIndexes
+    // and strings
+    LocalMemory<const char*> territoryTypes;
+    LocalMemory<int32_t> territoryReplacementIndexes;
+    readTerritoryAlias(territoryAlias.getAlias(),
+                       strings.getAlias(),
+                       territoryTypes,
+                       territoryReplacementIndexes,
+                       territoryLength, status);
+
+    // Read the variantAlias into variantTypes, variantReplacementIndexes
+    // and strings
+    LocalMemory<const char*> variantTypes;
+    LocalMemory<int32_t> variantReplacementIndexes;
+    readVariantAlias(variantAlias.getAlias(),
+                     strings.getAlias(),
+                     variantTypes,
+                     variantReplacementIndexes,
+                     variantLength, status);
+
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+
+    // We can only use strings after freeze it.
+    strings->freeze();
+
+    // Build the languageMap from languageTypes & languageReplacementIndexes
+    LocalPointer<CharStringMap> languageMap(new CharStringMap(490, status));
+    for (int32_t i = 0; U_SUCCESS(status) && i < languagesLength; i++) {
+        languageMap->put(languageTypes[i],
+                         strings->get(languageReplacementIndexes[i]),
+                         status);
+    }
+
+    // Build the scriptMap from scriptTypes & scriptReplacementIndexes
+    LocalPointer<CharStringMap> scriptMap(new CharStringMap(1, status));
+    for (int32_t i = 0; U_SUCCESS(status) && i < scriptLength; i++) {
+        scriptMap->put(scriptTypes[i],
+                       strings->get(scriptReplacementIndexes[i]),
+                       status);
+    }
+
+    // Build the territoryMap from territoryTypes & territoryReplacementIndexes
+    LocalPointer<CharStringMap> territoryMap(new CharStringMap(650, status));
+    for (int32_t i = 0; U_SUCCESS(status) && i < territoryLength; i++) {
+        territoryMap->put(territoryTypes[i],
+                          strings->get(territoryReplacementIndexes[i]),
+                          status);
+    }
+
+    // Build the variantMap from variantTypes & variantReplacementIndexes.
+    LocalPointer<CharStringMap> variantMap(new CharStringMap(2, status));
+    for (int32_t i = 0; U_SUCCESS(status) && i < variantLength; i++) {
+        variantMap->put(variantTypes[i],
+                        strings->get(variantReplacementIndexes[i]),
+                        status);
+    }
+
+    // copy hashtables
+    return new AliasData(
+        languageMap.orphan(),
+        scriptMap.orphan(),
+        territoryMap.orphan(),
+        variantMap.orphan(),
+        strings.orphan());
```

Comment:
... and then call `strings.orphanCharStrings()` and store the CharString, and then let the UniqueCharStrings object go out of scope.

(Like in loclikelysubtags.cpp)

Reply:
ok. Well do.